ocrd/ocrd/mets_server.py

```diff
@@ -0,0 +1,253 @@
+    """
+
+    def __init__(self, el, mimetype=None, pageId=None, loctype='OTHER', local_filename=None, mets=None, url=None, ID=None, fileGrp=None):
+        """
+        Args:
+            el (): ignored
+        Keyword Args:
+            mets (): ignored
+            mimetype (string): ``@MIMETYPE`` of this ``mets:file``
+            pageId (string): ``@ID`` of the physical ``mets:structMap`` entry corresponding to this ``mets:file``
+            loctype (string): ``@LOCTYPE`` of this ``mets:file``
+            local_filename (): ignored
+            url (string): ``@xlink:href`` of this ``mets:file``
+            ID (string): ``@ID`` of this ``mets:file``
+        """
+        self.ID = ID
+        self.mimetype = mimetype
+        self.url = url
+        self.loctype = loctype
+        self.pageId = pageId
+        self.fileGrp = fileGrp
+
+class ClientSideOcrdMets():
+    """
+    Replacement for :py:class:`ocrd_models.ocrd_mets.OcrdMets` with overrides for
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_files`,
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_all_files`, and
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.add_file` to query via HTTP a
+    :py:class:`ocrd.mets_server.OcrdMetsServer`.
+    """
+
+    def __init__(self, host, port, socket):
+        self.log = getLogger('ocrd.mets_client.%s' % ('uds' if socket else 'tcp'))
+        if socket:
+            self.url = f'http+unix://{socket.replace("/", "%2F")}'
+            self.session = requests_unixsocket_session()
+        else:
+            self.url = f'http://{host}:{port}'
+            self.session = requests_session()
+
+    def find_files(self, **kwargs):
+        r = self.session.request('GET', self.url, params={**kwargs})
+        for f in r.json()['files']:
+            yield ClientSideOcrdFile(None, ID=f['file_id'], pageId=f['page_id'], fileGrp=f['file_grp'], url=f['url'], mimetype=f['mimetype'])
+
+    def find_all_files(self, *args, **kwargs):
+        return list(self.find_files(*args, **kwargs))
+
+    def add_agent(self, *args, **kwargs):
+        return self.session.request('POST', f'{self.url}/agent', data=OcrdAgentModel.create(**kwargs))
+
+    @property
+    def file_groups(self):
+        return self.session.request('GET', f'{self.url}/file_groups').json()['file_groups']
+
+    @deprecated_alias(pageId="page_id")
+    @deprecated_alias(ID="file_id")
+    def add_file(self, file_grp, content=None, file_id=None, url=None, mimetype=None, page_id=None, **kwargs):
+        r = self.session.request(
+            'POST',
+            self.url,
+            data=OcrdFileModel.create(
+                file_id=file_id,
+                file_grp=file_grp,
+                page_id=page_id,
+                mimetype=mimetype,
+                url=url).dict(),
+            files={'data': content}
+        )
+
+    def save(self):
+        self.session.request('PUT', self.url)
+
+
+#
+# Server
+#
+
+class OcrdMetsServer():
```

Comment:
I personally prefer to extend FastAPI and then define the async functions as instance methods like in the processing-server-pr: https://github.com/OCR-D/core/pull/884/files#diff-a7b15eb188b2b6850fc8e727863dfd7fc8b9a676367990bc7129af7f5bb3950e
  . But I know that this has for example the drawback that the routes are not directly visible like in your version with the decorators and the routes have to be added with function calls.

Reply:
That is actually a nice pattern, I will look into it, thanks.

Reply:
I have decided against it, mostly because I find the solution as-is quite readable in comparison with programmatic adding of routes and also because we want to move ahead asap.